### PR TITLE
Create `workflow.Context` in cache

### PR DESCRIPTION
**Pull request by stephanos**
## What changed?

Start-Workflow and Signal-With-Start both write the new `workflow.Context` they create to the workflow cache.

## Why?

This change is required for Update-With-Start:
 - Update registry is stored inside of `workflow.Context`
 - Update-Workflow stores/retrieves the Update registry in/from the workflow cache
 - when Start-Workflow and Signal-With-Start create a new `workflow.Context`, they create it without writing to the workflow cache (see `NewWorkflowWithSignal`)
  - that means, Update-with-Start doesn't work since the Update registry won't be in the cache on the Matching callback to History

Therefore, Start-Workflow and Signal-with-Start will write a newly created `workflow.Context` to the cache now.

_PS: Note that the Update registry might move out of the cache in the future eventually_

_PPS: A larger change to store the MutableState in the cache, too, was considered and prototyped - but decided to leave out of scope_

## How did you test it?

Existing tests. (they were very successful in catching issues during development)

## Potential risks

After some discussion with Yichao and Alex, we don't think there's a risk here.

The only edge case we identified is when lots of _failed_ Workflow starts occur (ie due to the id reuse policy) and they flood the cache with abandoned `workflow.Context`s. But that seems unlikely, to us, to actually be a real problem.

### CODE REVIEW

File: service/history/api/create_workflow_util.go [hunk 1]
```diff
@@ -125,18 +126,17 @@ func NewWorkflowWithSignal(
...
+		shard,
+		namespaceEntry.ID(),
+		&commonpb.WorkflowExecution{WorkflowId: startRequest.StartRequest.WorkflowId, RunId: runID},
+		workflow.LockPriorityHigh,
 	)
-	return NewWorkflowLease(newWorkflowContext, wcache.NoopReleaseFn, newMutableState), nil
+	if err != nil {
+		return nil, err
+	}
+	return NewWorkflowLease(newWorkflowContext, releaseFn, newMutableState), nil
```
Comment (stephanos) [inline service/history/api/create_workflow_util.go:139]: This is the main change

File: service/history/api/create_workflow_util.go [hunk 2]
```diff
@@ -125,18 +126,17 @@ func NewWorkflowWithSignal(
...
+		shard,
+		namespaceEntry.ID(),
+		&commonpb.WorkflowExecution{WorkflowId: workflowID, RunId: runID},
+		workflow.LockPriorityHigh,
 	)
-	return NewWorkflowLease(newWorkflowContext, wcache.NoopReleaseFn, newMutableState), nil
+	if err != nil {
+		return nil, err
+	}
+	return NewWorkflowLease(newWorkflowContext, releaseFn, newMutableState), nil
```
Comment (stephanos) [inline service/history/api/create_workflow_util.go:139]: This is the main change

File: service/history/api/create_workflow_util.go [hunk 3]
```diff
@@ -125,18 +126,17 @@ func NewWorkflowWithSignal(
...
-		shard.GetConfig(),
-		definition.NewWorkflowKey(
-			namespaceEntry.ID().String(),
-			workflowID,
-			runID,
-		),
-		shard.GetLogger(),
-		shard.GetThrottledLogger(),
-		shard.GetMetricsHandler(),
+	newWorkflowContext, releaseFn, err := cache.GetOrCreateWorkflowExecution(
```
Comment (yycptt) [inline service/history/api/create_workflow_util.go:129]: nit: can consistency checker be used here? I am fine either way I guess, I just think we don't have a consistent way of using consistency checker and workflow cache in api handlers.

I thought the intention for consistency checker is for API handler implementations to use it instead of cache directly. But I don't know why consistency checker also allows retrieving workflow cache. so maybe my understanding is wrong.

File: service/history/workflow/context.go [hunk 4]
```diff
@@ -873,11 +875,12 @@ func (c *ContextImpl) ReapplyEvents(
...
 
-func (c *ContextImpl) UpdateRegistry(ctx context.Context) update.Registry {
+func (c *ContextImpl) UpdateRegistry(ctx context.Context, ms MutableState) update.Registry {
 	if c.updateRegistry == nil {
-		nsIDStr := c.MutableState.GetNamespaceEntry().ID().String()
+		// NOTE: cannot use Context's MutableState, as it might not be loaded - use the parameter instead.
+		nsIDStr := ms.GetNamespaceEntry().ID().String()
 		c.updateRegistry = update.NewRegistry(
-			func() update.Store { return c.MutableState },
+			func() update.Store { return ms },
```
Comment (yycptt) [inline service/history/workflow/context.go:883]: Not related to this PR. 

My understanding is with the change the loaded ms still may not be consistent with the currently cached update registry. And we need to resolve the lifecycle issue between those two components. Can you maybe add a TODO here? I think @alexshtin is working on this issue.

File: service/history/api/startworkflow/api.go [hunk 5]
```diff
@@ -359,58 +363,45 @@ func (s *Starter) resolveDuplicateWorkflowID(
...
-		definition.NewWorkflowKey(
-			s.namespace.ID().String(),
-			workflowID,
-			currentWorkflowConditionFailed.RunID,
-		),
-		prevExecutionUpdateAction,
+		currentExecutionWorkflowKey,
+		currentExecutionUpdateAction,
 		func() (workflow.Context, workflow.MutableState, error) {
-			workflowLease, err := api.NewWorkflowWithSignal(
```
Comment (yycptt) [inline service/history/api/startworkflow/api.go:381]: interesting, not sure why we create a new mutable state before.

Comment (yycptt) [reply service/history/api/startworkflow/api.go:381]: hmmm, I think it's because the transaction for newExecutionWorkflowLease has already been closed once above. And GetAndUpdateWorkflowWithNew will try to close it again. If it works, it works in a kinda weird way.

File: service/history/api/create_workflow_util.go (hunk 3 above)
Comment (stephanos) [reply service/history/api/create_workflow_util.go:129]: We definitely don't have a consistent way to do that, I agree.

I just tried it, the problem with using the ConsistencyChecker is that it only has a `GetWorkflowLease` method; not a `GetOrCreateWorkflowLease` method. So it won't work since it tries to load the Workflow from the database.

File: service/history/workflow/context.go (hunk 4 above)
Comment (stephanos) [reply service/history/workflow/context.go:883]: Right; this PR shouldn't change any of that behavior. For the existing use cases, the source of the MS is still the same (ie the `workflow.Context`). But for the case where the Context hasn't been persisted yet, it'll use the pass-in, in-memory MS instead of crashing due to a `nil` pointer exception.

File: service/history/api/startworkflow/api.go (hunk 5 above)
Comment (stephanos) [reply service/history/api/startworkflow/api.go:381]: Oh, I see. I didn't realize that wasn't idempotent. The reason I made the change is that it's competing for the same lock we already grabbed (as part of the cache write). 🤔

File: service/history/api/pollupdate/api_test.go [hunk 6]
```diff
@@ -68,8 +68,9 @@ type (
 
 	mockWorkflowLeaseCtx struct {
 		api.WorkflowLease
-		GetContextFn   func() workflow.Context
-		GetReleaseFnFn func() wcache.ReleaseCacheFunc
+		GetContextFn        func() workflow.Context
+		GetReleaseFnFn      func() wcache.ReleaseCacheFunc
+		GetUpdateRegistryFn func() update.Registry
```
Comment (alexshtin) [inline service/history/api/pollupdate/api_test.go:73]: You don't need this.

File: service/history/workflow/context.go (hunk 4 above)
Comment (stephanos) [reply service/history/workflow/context.go]: Talked to Alex and changed the approach.

**Comment by stephanos** [comment]
Closing since I believe attempting to save the MS again is a dead end. I'm going to approach this differently.
